cmdstanpy/utils/stancsv.py

```diff
@@ -129,245 +134,384 @@ def parse_hmc_adaptation_lines(
+
+    filtered = filter_csv_bytes_by_columns(sampling_draws, indexes_to_keep)
+    arr = csv_bytes_list_to_numpy(filtered).astype(int)
+
+    num_max_treedepth = np.sum(arr[:, 0] == max_treedepth)
+    num_divergences = np.sum(arr[:, 1])
+    return num_max_treedepth, num_divergences
+
+
+def is_sneaky_fixed_param(header: str) -> bool:
+    """Returns True if the header line indicates that the sampler
+    ran with the fixed_param sampler automatically, despite the
+    algorithm listed as 'hmc'.
+
+    See issue #805"""
+    num_dunder_cols = sum(col.endswith("__") for col in header.split(","))
+
+    return (num_dunder_cols < 7) and "lp__" in header
+
+
+def count_warmup_and_sampling_draws(
+    stan_csv: Union[str, os.PathLike, Iterator[bytes]],
+) -> Tuple[int, int]:
+    """Scans through a Stan CSV file to count the number of lines in the
+    warmup/sampling blocks to determine counts for warmup and sampling draws.
+    """
+
+    def determine_draw_counts(lines: Iterator[bytes]) -> Tuple[int, int]:
+        is_fixed_param = False
+        header_line_idx = None
+        adaptation_block_idx = None
+        sampling_block_idx = None
+        timing_block_idx = None
+        for i, line in enumerate(lines):
+            if header_line_idx is None:
+                if b"fixed_param" in line:
+                    is_fixed_param = True
+                if line.startswith(b"lp__"):
+                    header_line_idx = i
+                    if not is_fixed_param:
+                        is_fixed_param = is_sneaky_fixed_param(
+                            line.strip().decode()
+                        )
+                continue
+
+            if not is_fixed_param and adaptation_block_idx is None:
+                if line.startswith(b"#"):
+                    adaptation_block_idx = i
+            elif sampling_block_idx is None:
+                if not line.startswith(b"#"):
+                    sampling_block_idx = i
+                elif line.startswith(b"#  Elapsed"):
+                    sampling_block_idx = i
+                    timing_block_idx = i
+            elif timing_block_idx is None:
+                if line.startswith(b"#"):
+                    timing_block_idx = i
+            else:
+                break
+        else:
+            # Will raise if lines exhausts without all blocks being identified
+            raise ValueError(
+                "Unable to count warmup and sampling draws from Stan csv"
+            )
+
+        if is_fixed_param:
+            num_warmup = 0
+        else:
+            num_warmup = (
+                adaptation_block_idx - header_line_idx - 1  # type: ignore
+            )
+        num_sampling = timing_block_idx - sampling_block_idx
+        return num_warmup, num_sampling
+
+    if isinstance(stan_csv, (str, os.PathLike)):
+        with open(stan_csv, "rb") as f:
+            return determine_draw_counts(f)
+    else:
+        return determine_draw_counts(stan_csv)
```

Comment:
almost worth extracting this pattern out into a `call_on_file(func, stan_csv)`

Reply:
I personally fall on the side that repeating this short pattern twice isn't _quite_ enough to justify another function, but I don't hold that belief too strongly. Happy to change if you want.